### PR TITLE
[docs] [#9111] add cloud native dev with Gitpod and Codespaces

### PR DESCRIPTION
**Pull request by srinivasa-vasu**
A new content section added to show the code ready container-based development environment integration based on,
- GitHub Codespaces
- Gitpod workspaces

PR for #9111 

Updated section weightage to align the side menu with the base tile section order alignment

### CODE REVIEW

File: docs/content/latest/develop/gitdev/codespace.md [hunk 1]
```diff
@@ -23,8 +23,8 @@ Codespaces doesn't require anything in your local workstation other than a code
 ## Getting Started with a boot app
 You can find the source at [Spring Boot todo on GitHub](https://github.com/srinivasa-vasu/todo).
 
-- The easy way to get started with Codespaces is to simply fork this source repo and jump on to the [Setting up the Codespace environment](#setting-up-the-codespace-environment) section to launch the Codespaces environment for your forked repository. 
-- If you want to set this up from scratch, then follow the below instructions to bootstrap the base project template and copy the appropriate files/content from the [source repo]((https://github.com/srinivasa-vasu/todo)).
+- The easy way to get started with Codespaces is to simply fork this [source repo](https://github.com/srinivasa-vasu/todo) and jump on to the [Setting up the Codespace environment](#setting-up-the-codespace-environment) section to launch the Codespaces environment for your forked repository.
```
Comment (polarweasel) [inline docs/content/latest/develop/gitdev/codespace.md:26]: I really don't want to include stuff from anyone's personal GitHub repos. Can you please make a Yugabyte-owned repo for this, @srinivasa-vasu? (Jeff is the contact person for that)

File: docs/content/latest/develop/gitdev/gitpod.md [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    identifier: gitpod
+    parent: gitdev
+    weight: 592
+isTocNested: true
+showAsideToc: true
+---
+
+Use the [Gitpod](https://www.gitpod.io) workspaces to provision an instant development environment with a pre-configured YugabyteDB.
+
+Gitpod is a configurable ready to code cloud development environment accessible via a browser. A Gitpod workspace includes everything developers need to develop for a specific repository, including the `vscode` or `theia` editing experience, common languages, tools, and utilities. Instantly it sets up a cloud-hosted, containerized, and customizable vscode/theia environment.
```
Comment (polarweasel) [inline docs/content/latest/develop/gitdev/gitpod.md:16]: Theia is gone as of April: https://www.theregister.com/2021/04/08/gitpod_talks_up_importance_of/

```suggestion
Gitpod is a configurable ready-to-code cloud development environment accessible via a browser. A Gitpod workspace includes everything you need to develop for a specific repository, including the Visual Studio Code editing experience, common languages, tools, and utilities. Instantly it sets up a cloud-hosted, containerized, and customizable editing environment.
```

File: docs/content/latest/develop/gitdev/gitpod.md [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+Follow the steps below to set up a Gitpod workspace environment with a pre-configured YugabyteDB. For details on Gitpod workspaces, see the [Gitpod documentation](https://www.gitpod.io/docs/).
```
Comment (polarweasel) [inline docs/content/latest/develop/gitdev/gitpod.md:18]:
```suggestion
Follow the steps on this page to set up a Gitpod workspace environment with a pre-configured YugabyteDB. For details on Gitpod workspaces, refer to the [Gitpod documentation](https://www.gitpod.io/docs/).
```

**Review by polarweasel — CHANGES_REQUESTED** [review]
Have to stop for tonight, will try to finish reviewing tomorrow...

**Comment by polarweasel** [comment]
@aishwarya24 can you take an editing pass over this, and then we'll get this ready to merge?

**Review by aishwarya24 — COMMENTED** [review]
@srinivasa-vasu Not sure about this , but in the codespaces page , if there is a need to include the information about opening ports, please add them as well.

File: docs/content/latest/develop/gitdev/codespace.md (hunk 1 above)
Comment (srinivasa-vasu) [reply docs/content/latest/develop/gitdev/codespace.md:26]: ping @jharveysmith

File: docs/content/latest/develop/gitdev/codespace.md [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+If you run the `Create codespace` with the above-updated spec, the development environment gets provisioned with a running YugabyteDB instance.
+
+![install YugabyteDB](/images/develop/gitdev/codespace/install-yb.gif)
+
+{{< note title="Note" >}}
+📌 GitHub codespaces provisions a fully integrated cloud-native development environment with an automated port forwarding to develop, build, and test applications right from the browser tab.
+{{< /note >}}
+
+## ⌛ Summary
```
[the remark this answers is not in the document]
Comment (srinivasa-vasu) [reply docs/content/latest/develop/gitdev/codespace.md:130]: The graphic under this section summarizes in terms of what one would see if the integration is successful; just to reiterate the benefit of this. There is nothing from a Cleanup standpoint.

[the remark this answers is not in the document]
Comment (polarweasel) [reply docs/content/latest/develop/gitdev/codespace.md:130]: What I'm looking for in a Cleanup section is how to dispose of your environment. We shouldn't be leaving them without a way to get rid of stuff they've created...

File: docs/content/latest/develop/gitdev/codespace.md [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+### Setting up the Codespace environment
+As of writing this, Codespaces is still a beta feature. If this feature is enabled for your GitHub account, then the codespace environment is initialized at [GitHub Codespaces](https://github.com/codespaces).
+![initalize the codespace environment](/images/develop/gitdev/codespace/init-codespace.png)
+
+If you don't keep any codespace-specific spec in the source repo, then on clicking `Create codespace`, a default [dev environment](https://github.com/microsoft/vscode-dev-containers/tree/main/containers/codespaces-linux) gets provisioned with `codespaces-linux` container. It is a universal image with pre-built language-specific libraries and commonly used utilities. As you need a dev environment with an integrated YugabyteDB, you have to customize the universal image. If the default conventions provided by the codespaces are not sufficient, then you can provide our own configuration.
+
+To initialize the codespace environment, open the source code in a local `vscode` editor. Install the following extensions,
+- Remote - Containers
+- GitHub Codespaces
```
Comment (polarweasel) [inline docs/content/latest/develop/gitdev/codespace.md:54]:
```suggestion
To initialize the codespace environment, open the source code in a local `vscode` editor. Install the following extensions:

- Remote - Containers
- GitHub Codespaces
```

File: docs/content/latest/develop/gitdev/codespace.md [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    identifier: codespace
+    parent: gitdev
+    weight: 591
+isTocNested: true
+showAsideToc: true
+---
+
+Use the [GitHub Codespaces](https://github.com/features/codespaces) to provision an instant development environment with a pre-configured YugabyteDB.
+
+Codespaces is a configurable cloud development environment accessible via a browser or through a local `vscode` editor. A codespace includes everything developers need to develop for a specific repository, including the `vscode` editing experience, common languages, tools, and utilities. Instantly it sets up a cloud-hosted, containerized, and customizable vscode environment.
```
[the remark this answers is not in the document]
Comment (polarweasel) [reply docs/content/latest/develop/gitdev/codespace.md:16]: Fixed

File: docs/content/latest/develop/gitdev/codespace.md [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+
+Follow the steps below to set up a codespace environment with a pre-configured YugabyteDB. For details on GitHub Codespaces, see the [GitHub Codespaces documentation](https://docs.github.com/en/codespaces).
+
+## Requirements ⏳
```
[the remark this answers is not in the document]
Comment (polarweasel) [reply docs/content/latest/develop/gitdev/codespace.md:20]: Fixed

File: docs/content/latest/develop/gitdev/codespace.md [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+Codespaces doesn't require anything in your local workstation other than a code editor and git cli. Much of the development happens in the cloud through a web browser though you have the option to use vscode locally.
+
+## Getting Started with a boot app
```
[the remark this answers is not in the document]
Comment (polarweasel) [reply docs/content/latest/develop/gitdev/codespace.md:23]: Fixing all

File: docs/content/latest/develop/gitdev/codespace.md [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+ARG YB_VERSION
+ARG ROLE
+
+USER root
+
+RUN apt-get update && export DEBIAN_FRONTEND=noninteractive && \
+	apt-get install -y netcat --no-install-recommends
+
+RUN curl -sSLo ./yugabyte.tar.gz https://downloads.yugabyte.com/yugabyte-${YB_VERSION}-linux.tar.gz \
+	&& mkdir yugabyte \
```
[the remark this answers is not in the document]
Comment (polarweasel) [reply docs/content/latest/develop/gitdev/codespace.md:81]: Fixing all

File: docs/content/latest/develop/gitdev/codespace.md [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+    && mv ./yugabyte /usr/local/ \
+    && ln -s /usr/local/yugabyte/bin/yugabyted /usr/local/bin/yugabyted \
+    && ln -s /usr/local/yugabyte/bin/ysqlsh /usr/local/bin/ysqlsh \
+    && chmod +x /usr/local/bin/yugabyted \
+    && chmod +x /usr/local/bin/ysqlsh \
+    && rm ./yugabyte.tar.gz
+
+RUN mkdir -p /var/ybdp \
+	&& chown -R $ROLE:$ROLE /var/ybdp \
+	&& chown -R $ROLE:$ROLE /usr/local/yugabyte
```
[the remark this answers is not in the document]
Comment (polarweasel) [reply docs/content/latest/develop/gitdev/codespace.md:92]: Fixing all

**Comment by polarweasel** [comment]
@aishwarya24 I've made a bunch of edits. Your turn! Please grab what's left in my comments, and also make your own changes and comments. We're nearly done with this. :)

**Comment by srinivasa-vasu** [comment]
@aishwarya24 - It's already mentioned in the Dockerfile. That's good enough. Codespaces detects it when you launch the workspace environment.

File: docs/content/latest/develop/gitdev/gitpod.md [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+
+## Requirements
+
+Gitpod doesn't require anything in your local workstation other than a code editor and Git CLI. Much of the development happens in the cloud through a web browser.
+
+## Get started with a boot app
+
+You can find the source at [Spring Boot todo on GitHub](https://github.com/yugabyte/yb-todo-app.git). 
+
+The easy way to get started with Gitpod is to simply fork this [source repo](https://github.com/yugabyte/yb-todo-app.git) and initialize the Gitpod workspace environment by invoking `https://gitpod.io/#[REPO_URL]` in a browser window. Replace `[REPO_URL]` with your forked repository URL and you should already have been conneted to the Gitpod account before launching the browser URL.
```
Comment (polarweasel) [inline docs/content/latest/develop/gitdev/gitpod.md:28]: "...should already have been connected..." - does this mean I should be logged into my Gitpod account before I do this step?

Comment (polarweasel) [reply docs/content/latest/develop/gitdev/gitpod.md:28]: Also, if i'm doing the easy way, do i need to follow any of the instructions on this page other than "fork the repo, browse to gitpod.io/#repo_url"? I suspect we need to say "do this, then skip to [Customize the Gitpod environment]"

Comment (srinivasa-vasu) [reply docs/content/latest/develop/gitdev/gitpod.md]: > "...should already have been connected..." - does this mean I should be logged into my Gitpod account before I do this step?
Yes, u need to be logged in. If not, you would be taken through the login flow.


> Also, if i'm doing the easy way, do i need to follow any of the instructions on this page other than "fork the repo, browse to gitpod.io/#repo_url"? I suspect we need to say "do this, then skip to [Customize the Gitpod environment]"
Easy way => just fork and launch the URL

File: docs/content/latest/develop/gitdev/gitpod.md [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+
+Gitpod doesn't require anything in your local workstation other than a code editor and Git CLI. Much of the development happens in the cloud through a web browser.
+
+## Get started with a boot app
+
+You can find the source at [Spring Boot todo on GitHub](https://github.com/yugabyte/yb-todo-app.git). 
+
+**The easy way to get started** with Gitpod is to simply fork this [source repo](https://github.com/yugabyte/yb-todo-app.git) and initialize the Gitpod workspace environment by invoking `https://gitpod.io/#[REPO_URL]` in a browser window. Replace `[REPO_URL]` with your forked repository URL and you should already have been connected to the Gitpod account before launching the browser URL.
+
+If you want **to set this up from scratch**, use the following instructions to bootstrap the base project template and copy the appropriate files/content from the [source repo](https://github.com/yugabyte/yb-todo-app.git).
```
Comment (aishwarya24) [inline docs/content/latest/develop/gitdev/gitpod.md:30]:
```suggestion
If you want **to set the Spring Boot app from scratch**, use the following instructions to bootstrap the base project template and copy the appropriate files/content from the [source repo](https://github.com/yugabyte/yb-todo-app.git).
```

File: docs/content/latest/develop/gitdev/gitpod.md [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+
+### Initialize the base project structure
+
+Spring todo is a Java Spring Boot reactive app. However, the steps to go through the Gitpod experience are agnostic of the language/framework. A quick way to get started with a spring boot app is via the [Spring Initializer](https://start.spring.io). Generate the base project structure with Webflux, Flyway, and R2DBC dependencies.
+
+![Set up the base project abstract](/images/develop/gitdev/gitpod/init-sb.png)
+
+### Complete the CRUD APIs
+
+Complete the todo-service by copying the source and build files from the [source repo](https://github.com/yugabyte/yb-todo-app.git) to your repo to handle GET, POST, PUT, and DELETE API requests.
```
Comment (aishwarya24) [inline docs/content/latest/develop/gitdev/gitpod.md:40]:
```suggestion
Complete the todo-service by copying the source and build files from the [source repo](https://github.com/yugabyte/yb-todo-app.git) to your repo to handle GET, POST, PUT, and DELETE API requests. Once you have your project ready, make sure to create a repository on Github and add your project to this repository.
```

File: docs/content/latest/develop/gitdev/gitpod.md [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+
+To get started quickly, you can use the universal image [pre-built containers](https://www.gitpod.io/docs/quickstart) or a language-specific image. It can be further customized to fit your needs either by extending them or by creating a new one. A simple click provisions the entire development environment in the cloud with an integrated powerful Visual Studio Code editor. The entire config to set up the development environment lives in the same source code repository. Follow the steps in the next sections to set up your Gitpod environment.
+
+### Set up the Gitpod environment
+
+You initialize the Gitpod workspace environment for a specific repository by invoking `https://gitpod.io/#[REPO_URL]`.
+
+![Initalize the workspace environment](/images/develop/gitdev/gitpod/init-workspace.png)
+
+You can either use a universal image with pre-configured libraries and commonly used utilities or a language-specific image. Create the integrated YugabyteDB workspace environment by customizing the base universal image.
```
Comment (aishwarya24) [inline docs/content/latest/develop/gitdev/gitpod.md:58]:
```suggestion
You can either use a universal image with pre-configured libraries and commonly used utilities or a language-specific image. Create the integrated YugabyteDB workspace environment by customizing the base universal image. Use the editor in your gitpod environment to add these files directly and commit them back to your github repository.
```

File: docs/content/latest/develop/gitdev/gitpod.md [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+    onOpen: ignore
+  - port: 13000
+    onOpen: ignore
+  - port: 9042
+    onOpen: ignore
+  - port: 12000
+    onOpen: ignore    
+```
+
+Launch the workspaces environment with this updated spec to provision the development environment with a running YugabyteDB instance. It opens two terminals, one terminal to run the DB task, and the other to compile and run the boot app.
```
Comment (aishwarya24) [inline docs/content/latest/develop/gitdev/gitpod.md:158]:
```suggestion
Make sure to commit the above files to your Github repository from your existing gitpod environment. 
Next, [launch](../../../latest/develop/gitdev/gitpod/#set-up-the-gitpod-environment) the workspaces environment again with this updated spec to provision the development environment with a running YugabyteDB instance. It opens two terminals, one terminal to run the DB task, and the other to compile and run the boot app.
```

**Comment by aishwarya24** [comment]
Suggested few changes to make sure all the steps are explicitly mentioned.

**Review by polarweasel — APPROVED** [review]
Good to go. Once the preview builds, I'll merge this.

**Pull request merged**